src/app/clusters/thermostat-server/thermostat-server.cpp

```diff
@@ -1287,102 +1306,122 @@ bool emberAfThermostatClusterSetActivePresetRequestCallback(
-    if (gThermostatAttrAccess.GetPresetsEditable(endpoint))
-    {
-        ExtendTimer(endpoint, commandData.timeoutSeconds);
-        commandObj->AddStatus(commandPath, imcode::Success);
-        return true;
+        switch (attributeId)
+        {
+        case Presets::Id:
+            if (requestedPresets)
+            {
+                return false;
+            }
+            requestedPresets = true;
+            break;
+        case Schedules::Id:
+            if (requestedSchedules)
+            {
+                return false;
+            }
+            requestedSchedules = true;
+            break;
+        default:
+            return false;
+        }
     }
-
-    // Set presets editable to true and the scoped originator node id to the source scoped node id, and start a timer with the
-    // timeout in seconds passed in the command args. Return success.
-    gThermostatAttrAccess.SetPresetsEditable(endpoint, true);
-    gThermostatAttrAccess.SetOriginatorScopedNodeId(endpoint, sourceNodeId);
-    ScheduleTimer(endpoint, commandData.timeoutSeconds);
-    commandObj->AddStatus(commandPath, imcode::Success);
     return true;
 }
 
-bool emberAfThermostatClusterCancelPresetsSchedulesEditRequestCallback(
-    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
-    const chip::app::Clusters::Thermostat::Commands::CancelPresetsSchedulesEditRequest::DecodableType & commandData)
+bool handleAtomicBegin(chip::app::CommandHandler * commandObj, const ScopedNodeId & sourceNodeId,
+                       const chip::app::ConcreteCommandPath & commandPath,
+                       const chip::app::Clusters::Thermostat::Commands::AtomicRequest::DecodableType & commandData)
 {
     EndpointId endpoint = commandPath.mEndpointId;
 
-    // If presets are not editable, return INVALID_IN_STATE.
-    if (!gThermostatAttrAccess.GetPresetsEditable(endpoint))
+    if (!commandData.timeout.HasValue())
     {
-        commandObj->AddStatus(commandPath, imcode::InvalidInState);
+        commandObj->AddStatus(commandPath, imcode::InvalidCommand);
         return true;
     }
 
-    ScopedNodeId sourceNodeId = GetSourceScopedNodeId(commandObj);
+    auto timeout = commandData.timeout.Value();
 
-    // If the node id sending the CancelPresetsSchedulesRequest command is not the same as the one which send the
-    // previous StartPresetsSchedulesEditRequest, return UNSUPPORTED_ACCESS.
-    if (gThermostatAttrAccess.GetOriginatorScopedNodeId(endpoint) != sourceNodeId)
+    if (!validAtomicAttributes(commandData))
     {
-        commandObj->AddStatus(commandPath, imcode::UnsupportedAccess);
+        commandObj->AddStatus(commandPath, imcode::InvalidCommand);
         return true;
     }
 
-    Delegate * delegate = GetDelegate(endpoint);
-
-    if (delegate == nullptr)
+    bool inAtomicWrite = gThermostatAttrAccess.InAtomicWrite(endpoint);
+    if (inAtomicWrite)
```

Comment:
Per spec, if `sourceNodeId` is the thing doing the atomic write we should be returning INVALID_IN_STATE, no?

And that check comes before the "does not have a timeout" check.

How it's ordered with the validAtomicAttributes() check is not very clear, though the way the spec is laid out people would assume it comes after....

In general, since these checks return different status, the spec needs to define the order. Otherwise you might get disagreement between implementations on what happens if list of ids is invalid and the client is already associated with an atomic write state.